engine/access/rest/websockets/controller.go

```diff
@@ -129,23 +132,31 @@ func (c *Controller) configureKeepalive() error {
 // Parameters:
 // - ctx: Context to control cancellation and timeouts.
 func (c *Controller) monitorInactivity(ctx context.Context) error {
+	defer c.inactivityTracker.Stop()
+
 	for {
 		select {
 		case <-ctx.Done():
 			return nil
-		default:
+		case <-c.inactivityTracker.C():
```

Comment:
I think we can reduce the complexity a bit more.

Instead of precisely measuring the timeout, how about just waking up periodically to check how long the connection has been inactive? Maybe 1/10 of the inactivity timeout. this could even go in the `writeMessage` loop to avoid an additional goroutine.

I think it would be simplest to track inactivity based on when messages are sent. for example, when `writeMessages` is created, we could capture the current time as `lastMessageSentAt`. Each time a message is sent, we update the time.

Then, periodically, the loop checks if it's been more than `InactivityTimeout` seconds since `lastMessageSentAt`. If it has and` `c.dataProviders.Size() == 0`, then we close the connection.

This takes advantage of the fact that we send a message after unsubscribing, so we can easily track when the last provider was closed.

Reply:
[20c3e06](https://github.com/onflow/flow-go/pull/6817/commits/20c3e064b750c9672a7728da2440cd2f6484eb57)